mlir/lib/Analysis/Presburger/Coalesce.cpp

```diff
@@ -0,0 +1,909 @@
+
+// compute the non-pure adj_ineq case and return whether it has worked.
+// Constraint t is the adj_ineq.
+//
+// In the non-pure adj_ineq case, one of the polytopes is like an extension of
+// the other one. This can be computed, by inverting the adj_ineq and checking
+// whether all constraints are stll valid for this new polytope.
+//   ____          ____
+//  |    \        |    \
+//  |     \   ==> |     \
+//  |_____|_      |______\
+//
+bool adjIneqCase(SmallVector<FlatAffineConstraints, 4> &basicSetVector,
+                 unsigned i, unsigned j, const Info &info_a,
+                 const Info &info_b);
+
+// compute the pure adjEqCase and return whether it has worked.
+//
+// The pure case consists of two equalities, that are adjacent. Such equalities
+// can always be coalesced by finding the two constraints, that make them become
+// a trapezoid. This is done by wrapping the limiting constraints of the
+// equalities around each other to include the other one.
+//                _
+//    / /        / /
+//   / /   ==>  / /
+//  / /        /_/
+//
+bool adjEqCasePure(SmallVectorImpl<FlatAffineConstraints> &basicSetVector,
+                   unsigned i, unsigned j, const Info &info_a,
+                   const Info &info_b);
+
+// compute the adj_eq Case for no CUT constraints.
+//
+// The adj_eq no cut case is simply an extension case, where the constraint,
+// that is adjacent to the equality, can be relaxed by 1 to include the other
+// polytope.
+//   ______           _______
+//  |      ||        |       |
+//  |      ||        |       |
+//  |      ||   ==>  |       |
+//  |______||        |_______|
+//
+bool adjEqCaseNoCut(SmallVectorImpl<FlatAffineConstraints> &basicSetVector,
+                    unsigned i, unsigned j, SmallVector<int64_t, 8> t);
+
+// compute the non-pure adjEqCase and return whether it has worked.
+//
+// The non-pure case has cut constraints such that it is not a simple extension
+// like the no cut case. It is computed by wrapping those cut constraints and
+// checking, whether everything stilly holds.
+//     ________           ________
+//    |        |         |        \
+//    |        | |  ==>  |         |
+//    |        |         |        /
+//    |        /         |       /
+//    |_______/          |______/
+//
+bool adjEqCaseNonPure(SmallVectorImpl<FlatAffineConstraints> &basicSetVector,
+                      unsigned i, unsigned j, const Info &info_a,
+                      const Info &info_b);
+
+void shift(SmallVectorImpl<int64_t> &t, int amount) {
+  t.push_back(t.pop_back_val() + amount);
+}
+
+SmallVector<int64_t, 8> complement(ArrayRef<int64_t> t) {
+  SmallVector<int64_t, 8> complement;
+  for (size_t k = 0; k < t.size() - 1; k++) {
+    complement.push_back(-t[k]);
+  }
+  complement.push_back(-t.back() - 1);
+  return complement;
+}
+
+// helperfuncton to convert arrayRefs to SmallVectors
+static SmallVector<int64_t, 8> arrayRefToSmallVector(ArrayRef<int64_t> ref) {
```

Comment:
You can use `llvm::to_vector`.

Reply:
How do you mean this? I tried substituting this fucntion with llvm::to_vector, but this seems to take a size parameter and not an `ArrayRef.`

Reply:
`to_vector` takes a `Range`. Note that this isn't per se a specific class as they extract begin and end with `std::begin` and `std::end`. 

https://llvm.org/doxygen/SmallVector_8h_source.html line 964

It seems like it also requires a `Size` template parameter that in you case would be 8.

Reply:
I understood this now and use `llvm::to_vector`for this.